Core/HttpHandlerProxy.cs

```diff
@@ -0,0 +1,151 @@
+        /// </summary>
+        /// <returns>true to send a WWW-authenticate HTTP header with requests after authentication has taken place; otherwise, false. The default is false.</returns>
+        public bool PreAuthenticate { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="T:System.Boolean"/> value that controls whether default credentials are sent with requests.
+        /// </summary>
+        /// <returns>true if the default credentials are used; otherwise false. The default value is false.</returns>
+        public bool UseDefaultCredentials { get; set; }
+
+        public HttpHandlerProxy(IHttpMessageHandlerFactory httpMessageHandlerFactory, string httpClientName)
+        {
+            this.httpClientName = httpClientName;
+            this.httpMessageHandler = httpMessageHandlerFactory?.CreateHandler(httpClientName);
+            this.httpClientHandler = null == httpMessageHandlerFactory ? HttpClientExtensions.CreateHttpClientHandler(httpClientName) : null;
+        }
+
+        private void Authorize(HttpRequestMessage request)
+        {
+            switch (Credentials)
+            {
+                case NetworkCredential networkCredential:
+                    {
+                        string authString = !string.IsNullOrEmpty(networkCredential.Domain)
+                            ? networkCredential.Domain + "\\" + networkCredential.UserName + ":" + networkCredential.Password
+                            : networkCredential.UserName + ":" + networkCredential.Password;
+
+                        string base64AuthString = Convert.ToBase64String(Encoding.UTF8.GetBytes(authString));
+
+                        request.Headers.Authorization = new AuthenticationHeaderValue("Basic", base64AuthString);
+                    }
+                    break;
+
+                case null:
+                    break;
+
+                default:
+                    throw new AuthenticationException($"Credentials of type '{Credentials.GetType().Name}' is not supported");
+            }
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var cookieHeader = CookieContainer?.GetCookieHeader(request.RequestUri);
+            if (!string.IsNullOrEmpty(cookieHeader))
+            {
+                request.Headers.Add(HeaderNames.Cookie, cookieHeader);
+            }
+
+            Authorize(request);
+
+            HttpResponseMessage response = null;
+            if (null != InnerHandler)
+            {
+                response = await base.SendAsync(request, cancellationToken);
+            }
+            else if (null != httpMessageHandler)
+            {
+                var delegatingHandler = new SimpleHttpMessageHandlerProxy(httpMessageHandler);
+                response = await delegatingHandler.SendImplAsync(request, cancellationToken);

```

Comment:
What is the goal with SendImplAsync and SimpleHttpMessageHandlerProxy? I think you can drop it

Reply:
Main problem here is that after you made any call through InnerHandler you can't change InnerHandler. In case of negotiate I need to make a call and create new one handler.

Reply:
I still not clear why this won't work. You do same thing with httpMessageHandler and drop delegatingHandler at all
```
response = await httpMessageHandler.SendAsync(request, cancellationToken);
```

Reply:
you can't call httpMessageHandler.SendAsync directlly only through custom delegating handler with public custom send method

Reply:
Got it, I missed the fact it's protected in definition